src/client/ua_client_connect.c

```diff
@@ -1513,6 +1557,128 @@ UA_Client_connectSecureChannel(UA_Client *client, const char *endpointUrl) {
     return client->connectStatus;
 }
 
+static void
+__Client_reverseConnectCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                         void *application, void **connectionContext,
+                         UA_ConnectionState state, const UA_KeyValueMap *params,
+                         UA_ByteString msg) {
+    UA_Client *client = (UA_Client*)application;
+
+    UA_LOCK(&client->clientMutex);
+
+    if (!client->reverseConnectListenConnectionId && !client->listenerDisconnected) {
+        client->reverseConnectListenConnectionId = connectionId;
+    }
+
+    if (connectionId == client->reverseConnectListenConnectionId) {
+        client->reverseConnectListeningConnectionState = state;
+    }
+
+    if (connectionId == client->reverseConnectListenConnectionId && state == UA_CONNECTIONSTATE_CLOSING) {
+        client->reverseConnectListenConnectionId = 0;
+        UA_UNLOCK(&client->clientMutex);
+        return;
+    }
+
+    if (connectionId != client->reverseConnectListenConnectionId && !client->listenerDisconnected) {
+        cm->closeConnection(cm, client->reverseConnectListenConnectionId);
+        client->listenerDisconnected = true;
+    }
+
+    if (connectionId != client->reverseConnectListenConnectionId) {
+        UA_UNLOCK(&client->clientMutex);
+        __Client_networkCallback(cm, connectionId, application, connectionContext, state, params, msg);
+        return;
+    }
+
+    UA_UNLOCK(&client->clientMutex);
+}
+
+UA_StatusCode UA_Client_startListeningForReverseConnect(UA_Client *client, const UA_String *listenHostnames,
+                                                        size_t listenHostnamesLength,
+                                                        UA_UInt16 port) {
+    UA_Client_disconnect(client);
```

Comment:
Don't disconnect here.
Rather check if the client is already connected.
And abort in that case.
Similar to the "normal" connect.